containers/src/Data/Tree.hs

```diff
@@ -181,14 +182,55 @@ mfixTree f
 instance Traversable Tree where
     traverse f (Node x ts) = liftA2 Node (f x) (traverse (traverse f) ts)
 
+-- | Folds in preorder
+
+-- See Note [Implemented Foldable Tree functions]
 instance Foldable Tree where
-    foldMap f (Node x ts) = f x `mappend` foldMap (foldMap f) ts
+    fold = foldMap id
+    {-# INLINABLE fold #-}
+
+    foldMap f = go
+      where go (Node x ts) = f x `mappend` foldMap go ts
+    {-# INLINE foldMap #-}
+
+    foldr f z = \t -> go t z  -- Use a lambda to allow inlining with two arguments
+      where
+        go (Node x ts) = f x . foldr ((.) . go) id ts
```

Comment:
In https://gitlab.haskell.org/ghc/ghc/-/issues/22659#note_469510 SPJ identifies why this is faster. And sebastian suggest that this variant may be slower for deep trees (i.e. something like `Node 1 $ singleton $ Node 2 $ singleton $ ...`)

I suggest:
- Add benchmarks for "simple" foldr as well, so the difference can be checked later as well
- Add benchmark for deep trees
- Rewrite this using two mutually recursive functions as SPJ shows. That implementation is less magical, relies less on optimizer, and is easy to explain why it behaves better.

Reply:
> sebastian suggest that this variant may be slower for deep trees

That is not what Sebastian is saying. If I understand the comment correctly, Sebastian says that if we had two mutually recursive functions after optimization, we would allocate more. The current code optimizes to a single function.

> Add benchmarks for "simple" foldr as well, so the difference can be checked later as well

I don't see the point of this. Anyone running benchmarks would have to filter out the noise of a benchmark that is not live code. If anyone wants to test the simple version that's easy, they just need to comment out this line and uncomment the definition below.

> Add benchmark for deep trees

I have, please read the comment above.

> Rewrite this using two mutually recursive functions as SPJ shows. That implementation is less magical, relies less on the optimizer, and is easy to explain why it behaves better.

The second function out of SPJ's two mutually recursive functions manually case matches on the list and recurses. We don't need to write code like that when we have folds to do the job for us. And if we use foldr, that is the same as the simple definition, only broken into two, and it optimizes to the same inefficient code.

I honestly don't understand your dislike for this implementation.

Reply:
At the very least expand

```haskell
(.) . go
```

it's a pointless point-free form.

Reply:
> I don't see the point of this. Anyone running benchmarks would have to filter out the noise of a benchmark that is not live code. If anyone wants to test the simple version that's easy, they just need to comment out this line and uncomment the definition below.

I disagree about "easy". But as I'm not a maintainer, I cannot require you to add it either.

Reply:
> At the very least expand `(.) . go`

Sure thing, done.

Reply:
Ok now I undertstand what happens.

```haskell
foldr f z = \(Node x ts) -> f x (foldrForest f ts z)

foldrForest :: (a -> b -> b) -> [Tree a] -> b -> b
foldrForest f = go where 
  go []               z = z
  go (Node x xs : ys) z = f x (go xs (go ys z))
{-# INLINE foldrForest #-}
```

I find this easier to understand that one written using list's `foldr`. YMMV

Reply:
Yes, your function seems to end up in the same form as the current implementation. But I do not find the current one hard to understand. Of course, this is subjective. When something can be written as a fold, I would prefer to write it as a fold instead of explicit recursion unless there is a strong reason to do otherwise.

@treeowl do you have a preference here?
Otherwise, if there's nothing else to add, can we close this?